### ISSUE TITLE
[Feature Request] Server Side Render Routes

### ISSUE THREAD

**Issue by xXAvoraXx** [umijs/umi#12541]
## Background

我喜欢 UmiJs 的优点，尤其是插件逻辑，但我在开发插件方面有困难，下面是我试图公开开发的插件。

众所周知，路由是在 config.ts 中静态定义的，但在我看来这是一个安全漏洞，一些警惕性高的用户可以篡改源代码，看到不应该知道的路由。

我想从服务器上获取信息，了解用户登录后能看到哪些路由，不能看到哪些，然后对传入的路由进行渲染和打包。在这一点上，这个插件发挥了作用，它在我测试非插件版本时可以正常工作，但我需要帮助来改进这个插件。

I love the advantages of UmiJs, especially the plugin logic, but I have a hard time developing plugins, so below you can find the plugin I am trying to develop publicly.

Briefly what I'm trying to do in this plugin is this, as it is known, routes are statically defined in config.ts, but this is a security vulnerability in my opinion, some vigilant users can tamper with the source code and see routes that they should not know about.

I want to get the information from the server about which routes users can and cannot see after logging in, then the routes should be rendered and packaged. That's where this plugin comes in, it works when I test the non-plugin version but I need help to improve it.

## Proposal

https://github.com/xXAvoraXx/umi-plugin-ssr-routes

**Comment by xiaohuoni** [comment]
Although I don't quite understand your description, I think what you may need it
```ts
   api.addRuntimePlugin(() => {
      return [withTmpPath({ api, path: `${DIR_NAME}/runtime.tsx`, })];
    });
```

**Comment by AoaoMH** [comment]
umi自带的patchRoutes/patchClientRoutes似乎可以实现这个功能？

**Comment by xiaohuoni** [comment]
关于安全性，我觉得应该是在路由权限和接口权限上处理

**Comment by xXAvoraXx** [comment]
> umi自带的patchRoutes/patchClientRoutes似乎可以实现这个功能？

Yes, I explained how it should be done in the discussion here. After getting the routes from the api, the rendering process takes place and is matched with the menu data.
是的，我在这里的讨论中解释了应该如何操作。 从 api 获取路由后，将执行渲染过程并与菜单数据匹配。

https://github.com/ant-design/ant-design-pro/discussions/10898

**Comment by xXAvoraXx** [comment]
@AoaoMH you can try this -> https://www.npmjs.com/package/umi-plugin-ssr-routes

@xiaohuoni 
```
export default defineConfig({  
    ssrRoutes: {}
});



// app.tsx
export async function getServerSideRoutes(): Promise<RouteRaw[]> {
  return await getUserRoutes().then((res) => {
    if (res.data) {
      return res.data;
    } else {
      return [];
    }
  });
}

/** GET /api/getUserRoutes */
export async function getUserRoutes(options?: { [key: string]: any }) {
  return request<{
    data: RouteRaw[];
  }>('/api/GetUserRoutes', {
    method: 'GET',
    prefix: undefined,
    params: {
      clientId: 'uis',
    },
    ...(options || {}),
  });
}

```
